src/potentials.cpp

```diff
@@ -504,92 +504,171 @@ void FunctorPotential::from_json(const json &j) {
-                    else if (it->is_object())
-                        rmax2 = std::pow(it->at("default").get<double>(), 2);
+/**
+ * For each pair of atom types a file is created containing
+ * the exact and splined pair potential as a function of distance
+ */
+void TabulatedPotential::save_potentials() {
+    for (size_t i = 0; i < Faunus::atoms.size(); ++i) { // loop over atom types
+        for (size_t j = 0; j <= i; ++j) {               // and build matrix of spline data (knots) for each pair
+            if (atoms[i].implicit || atoms[j].implicit) {
+                continue;
+            }
+            Particle particle1 = Faunus::atoms.at(i);
+            Particle particle2 = Faunus::atoms.at(j);
+            auto filename = fmt::format("{}-{}_tabulated.dat", Faunus::atoms[i].name, Faunus::atoms[j].name);
+            if (std::ofstream stream(filename); stream) {
+                stream << "# r u_splined/kT u_exact/kT\n";
+                double rmax = sqrt(matrix_of_knots(i, j).rmax2);
+                for (double r = dr; r < rmax; r += dr) {
+                    stream << fmt::format("{:.6E} {:.6E} {:.6E}\n",
+                                          r, operator()(particle1, particle2, r *r, {r, 0, 0}),
+                                          TabulatedPotential::operator()(particle1, particle2, r *r, {r, 0, 0}));
                 }
+            }
+        }
+    }
+}
 
-                // adjust lower splining distance to match
-                // the given energy threshold (u_at_min2)
-                double dr = 1e-2;
-                while (rmin2 >= dr) {
-                    double u = std::fabs(this->umatrix(i, k)(a, b, rmin2, {0, 0, sqrt(rmin2)}));
-                    if (u > u_at_rmin * 1.1)
-                        rmin2 = rmin2 + dr;
-                    else if (u < u_at_rmin / 1.1)
-                        rmin2 = rmin2 - dr;
-                    else
-                        break;
-                }
+/**
+ * @param i Atom type index
+ * @param j Atom type index
+ * @param energy_threshold Absolute maximum energy used to determine min. distance for spline
+ * @param rmin Initial starting point for rmin
+ * @return Minimum distance for splining
+ *
+ * For repulsive potentials (at short sep.), the threshold is the maximum energy splined
+ * For attactive potentials (at short sep.), the threshold is the minimum energy splined
+ */
+double TabulatedPotential::findLowerDistance(int i, int j, double energy_threshold, double rmin) {
+    assert(rmin > 0);
+    Particle particle1 = Faunus::atoms.at(i);
+    Particle particle2 = Faunus::atoms.at(j);
+    int num_iterations = 0;
+    while (rmin >= dr) {
+        if (num_iterations++ == max_iterations) {
+            throw std::runtime_error("Pair potential spline error: cannot determine minimum distance");
+        }
+        double u = std::fabs(FunctorPotential::operator()(particle1, particle2, rmin *rmin, {rmin, 0, 0}));
```

Comment:
Yes, it's annoyingly the way clang-format wants it. Not sure there's a point in changing it as it will attack it again later. Best thing would be to fix it in the clang-format config file, but it doesn't look inviting.